File: steelpipes/sp_payment_entry/sp_payment_entry.js
frappe.ui.form.on('Payment Entry', {
    party: function(frm) {
        if (frm.doc.party) {
            frappe.call('steelpipes.sp_payment_entry.sp_payment_entry.get_contract',
            {party_type: frm.doc.party_type, party: frm.doc.party})
            .then (r => {
                if (r.message) {
                    frm.set_value('contact_person', r.message)
                }
            })
        }
    },

    contact_person: function(frm) {
        if (frm.doc.contact_person) {
            frappe.call('steelpipes.sp_payment_entry.sp_payment_entry.set_phone',
            {contact: frm.doc.contact_person})
            .then (r => {
                if (r.message) {
                    frm.set_value('contact_phone', r.message)
                    frm.refresh_fields('contact_phone')
                }
            })
        }
    }
});

File: steelpipes/sp_payment_entry/sp_payment_entry.py
import frappe
from frappe import _

@frappe.whitelist()
def set_phone(contact):
    doc = frappe.get_doc('Contact', contact)

    phone = []
    if doc.phone:
        return doc.phone
    elif doc.mobile_no:
        return doc.mobile_no
    elif len(doc.phone_nos) == 0:
        return False
    else:
        phone = [phone.phone for phone in doc.phone_nos]
        if phone[0]:
            return phone[0]
        else:
            return False
    return False

@frappe.whitelist()
def get_contract(party_type, party):
    contract = frappe.db.get_all('Dynamic Link', {
        'link_doctype': party_type,
        'link_name': party,
        'parenttype': 'Contact'
    }, 'parent')
    if contract:
        return contract[0].parent
    else:
        return False